### PR TITLE
Add CircleCI template

### LINKED ISSUE #1: CircleCI template
The script in [coq-ext-lib](//github.com/coq-community/coq-ext-lib/blob/master/.circleci/config.yml) is quite generic. Is it worth a template here?

### PR DESCRIPTION
**Pull request by liyishuai**
Resolve #1 
## Todo
- [x] Test compatibility with existing projects
- [x] Update `generate.sh`

### CODE REVIEW

**Comment by Zimmi48** [comment]
It seems to me that there are differences with the other CI templates that should not be specific to CircleCI like the use of `extra-dev` and the coinstability testing. Should we also add these to the other templates? cc @erikmd

**Comment by erikmd** [comment]
Also, as mentioned in the original post, the script `generate.sh` should be updated accordingly

File: config.yml.mustache [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    environment:
+      OPAMJOBS: 2
+      OPAMVERBOSE: 1
+      OPAMYES: true
+      TERM: xterm
+    steps:
+    - checkout
+    - run:
+        name: Pull submodules
+        command: git submodule update --init --remote
```
Comment (Zimmi48) [inline config.yml.mustache:20]: Is this yet another subtle difference between this template and the other CI templates?

File: config.yml.mustache [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+          opam repo -a --set-default add coq-extra-dev https://coq.inria.fr/opam/extra-dev
+          opam update
+          opam install --deps-only .
+    - run:
+        name: List installed packages
+        command: opam list
+    - run:
+        name: Build, test, and install package
+        command: opam install -t .
+{{# ci_test_dependants }}
```
Comment (Zimmi48) [inline config.yml.mustache:36]: Should be added to `ref.yml`.

File: config.yml.mustache [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    - run:
+        name: Test dependants
+        command: |
+          PINS=$(echo `opam list -s --pinned --columns=package` | sed 's/ /,/g')
```
Comment (erikmd) [inline config.yml.mustache:40]: The <code>$(echo \`opam list…\` | sed 's/ /,/g')</code> phrasing is OK, but maybe a bit hacky ;)
maybe we could have a sed-free solution with some array + IFS?

```
PINS=($(opam list -s --pinned --columns=package)); PINS=$(IFS=, ; echo "${PINS[*]}")
```
but this requires bash (inspired by https://stackoverflow.com/a/9429887/9164010)

File: config.yml.mustache [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+          PACKAGES=`opam list -s --depends-on {{ opam_name }}{{^ opam_name }}coq-{{ shortname }}{{/ opam_name }} --coinstallable-with $PINS`
+          if ! [ -z "$PACKAGES" ]
```
Comment (erikmd) [inline config.yml.mustache:42]: you might want to simplify `! [ -z "$PACKAGES" ]` to `[ -n "$PACKAGES"]` (?)

File: config.yml.mustache (hunk 1 above)
Comment (erikmd) [reply config.yml.mustache:20]: I think so… Maybe adding this `git submodule` part is unneeded for the default coq-community templates? at least for the time being (IMHO the three CI templates should be as close as possible semantically one from the other, and later on other PRs could just as well add features that enhance them, in a row?)

Comment (liyishuai) [reply config.yml.mustache:20]: Travis CI [clones Git submodules by default](//docs.travis-ci.com/user/customizing-the-build#git-submodules). I've added submodule for GitHub Actions.

File: config.yml.mustache (hunk 3 above)
Comment (erikmd) [reply config.yml.mustache]: Or maybe this is even simpler (no sed, only one subshell):

```
PINS=$(opam list -s --pinned --columns=package | xargs | tr ' ' ,)
```

**Comment by liyishuai** [comment]
> @liyishuai do you plan to add a similar script in Travis / GitHub Actions templates? or in a later PR.

Added to Travis.
For GitHub Actions, we need to:
1. Add option for testing dependants in https://github.com/coq-community/docker-coq-action
2. Configure the option via `ci_test_dependants`.

**Comment by erikmd** [comment]
Hi @liyishuai 
>  For GitHub Actions, we need to:
> 1. Add option for testing dependants in https://github.com/coq-community/docker-coq-action
> 2. Configure the option via ci_test_dependants.

Good point. So this can be done in a later PR maybe, indeed

File: config.yml.mustache (hunk 1 above)
Comment (erikmd) [reply config.yml.mustache:20]: OK! have you checked that this is exactly that command triggered by Travis for git submodule cloning?

(I'm not very submodule-savvy, but e.g. this blog https://mgalgs.github.io/2014/07/03/git-submodule-update-init-remote-recursive-considered-harmful.html mentions the following commands: `git submodule update --init --recursive; git submodule update --remote`)

Comment (liyishuai) [reply config.yml.mustache:20]: Travis uses `git submodule update --init --recursive`. Should I just drop `--remote`?

Comment (erikmd) [reply config.yml.mustache:20]: Maybe yes − this option is documented here:
https://git-scm.com/docs/git-submodule#Documentation/git-submodule.txt---remote ;
it appears to be related to remote-tracking branches (which would be a bit off-topic in the context of CI builds)

Comment (erikmd) [reply config.yml.mustache:20]: and this complies with the doc of GitLab CI as well:
https://docs.gitlab.com/ee/ci/git_submodules.html#using-git-submodules-in-your-ci-jobs
so `git submodule update --init --recursive` seems the way to go

Comment (erikmd) [reply config.yml.mustache:20]: (actually the GItLab CI page I mentioned suggests two lines)
```
git submodule sync --recursive
git submodule update --init --recursive
```

**Pull request merged**